Runtime/Model/Breadcrumbs/Storage/BacktraceStorageLogManager.cs

```diff
@@ -0,0 +1,309 @@
+
+        /// <summary>
+        /// Lock object
+        /// </summary>
+        private object _lockObject = new object();
+
+        /// <summary>
+        /// Current breadcurmbs fle size
+        /// </summary>
+        private long currentSize = 0;
+
+
+        /// <summary>
+        /// Queue that represents number of bytes in each log stored in the breadcrumb file
+        /// </summary>
+        private readonly Queue<long> _logSize = new Queue<long>();
+
+        internal IBreadcrumbFile BreadcrumbFile { get; set; }
+
+        public BacktraceStorageLogManager(string storagePath)
+        {
+            if (string.IsNullOrEmpty(storagePath))
+            {
+                throw new ArgumentException("Breadcrumbs storage path is null or empty");
+            }
+            BreadcrumbsFilePath = Path.Combine(storagePath, BreadcrumbLogFileName);
+            BreadcrumbFile = new BreadcrumbFile(BreadcrumbsFilePath);
+        }
+
+        /// <summary>
+        /// Enables breadcrumbs integration
+        /// </summary>
+        /// <returns>true if breadcrumbs file was created. Otherwise false.</returns>
+        public bool Enable()
+        {
+            try
+            {
+                if (BreadcrumbFile.Exists())
+                {
+                    BreadcrumbFile.Delete();
+                }
+
+                using (var _breadcrumbStream = BreadcrumbFile.GetCreateStream())
+                {
+                    _breadcrumbStream.Write(StartOfDocument, 0, StartOfDocument.Length);
+                    _breadcrumbStream.Write(EndOfDocument, 0, EndOfDocument.Length);
+                }
+                currentSize = StartOfDocument.Length + EndOfDocument.Length;
+            }
+            catch (Exception e)
+            {
+                System.Diagnostics.Debug.WriteLine(string.Format("Cannot initialize breadcrumbs file. Reason: {0}", e.Message));
+                return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Adds breadcrumb entry to the breadcrumbs file.
+        /// </summary>
+        /// <param name="message">Breadcrumb message</param>
+        /// <param name="level">Breadcrumb level</param>
+        /// <param name="type">Breadcrumb type</param>
+        /// <param name="attributes">Breadcrumb attributs</param>
+        /// <returns>True if breadcrumb was stored in the breadcrumbs file. Otherwise false.</returns>
+        public bool Add(string message, BreadcrumbLevel level, UnityEngineLogLevel type, IDictionary<string, string> attributes)
+        {
+            byte[] bytes;
+            lock (_lockObject)
+            {
+                long id = _breadcrumbId++;
+                var jsonObject = CreateBreadcrumbJson(id, message, level, type, attributes);
+                bytes = System.Text.Encoding.UTF8.GetBytes(jsonObject.ToJson());
+
+                if (currentSize + bytes.Length > BreadcrumbsSize)
```

Comment:
I think we only need to lock this part, we can save a lot of time by avoiding unnecessary locking, especially on every `Add` call

Reply:
I had the same feeling but on the other hand, what should we do when we're in the middle of cleaning breadcrumbs? We will remove the new breadcrumb, right?